### ISSUE TITLE
LocalRateLimiter with dynamic descriptor values 

### ISSUE THREAD

**Issue by kiran-bjn** [envoyproxy/envoy#19895]
*Description*:
> I have setup a localRateLimiter Filter on the route level . I am trying to configure the rateLimiter such that the limits are applied based on the remote-address or an Auth token passed in the header. Eg: For a particular API path /foo, Allow 5 reqs per min per token or remoteAddress. despite digging through the documentation multiple times, I am not able to find the relevant configuration to achieve this.  

LocalRateLimitConfig

`  - applyTo: HTTP_ROUTE
      match:
        context: SIDECAR_INBOUND
        routeConfiguration:
          vhost:
            name: inbound|http|80
      patch:
        operation: MERGE
        value:
          route:
            rate_limits:
              - actions:
                  - request_headers:
                      descriptor_key: AuthToken
                      header_name: Authorization
              - actions:
                  - header_value_match:
                      descriptor_value: quote-path-auth
                      expect_match: false
                      headers:
                        - name: Authorization
          typed_per_filter_config:
            envoy.filters.http.local_ratelimit:
              '@type': type.googleapis.com/udpa.type.v1.TypedStruct
              type_url: >-
                type.googleapis.com/envoy.extensions.filters.http.local_ratelimit.v3.LocalRateLimit
              value:
                descriptors:
                  - entries:
                      - key: AuthToken
                   //   - value: x-auth-xxx
                   //     -rate_limit:
                   //            requests_per_unit: 10
                   //            unit: minute  
                    token_bucket:
                      fill_interval: 20s
                      max_tokens: 2
                      tokens_per_fill: 2
                  - entries:
                      - key: header_match
                        value: quote-path-auth
                    token_bucket:
                      fill_interval: 20s
                      max_tokens:5
                      tokens_per_fill: 5
                filter_enabled:
                  default_value:
                    denominator: HUNDRED
                    numerator: 100
                  runtime_key: local_rate_limit_enabled
                filter_enforced:
                  default_value:
                    denominator: HUNDRED
                    numerator: 100
                  runtime_key: local_rate_limit_enforced
                local_rate_limit_per_downstream_connection: false
                response_headers_to_add:
                  - append: false
                    header:
                      key: x-local-rate-limit
                      value: 'true'
                stat_prefix: http_local_rate_limiter
                token_bucket:
                  fill_interval: 20s
                  max_tokens: 1000
                  tokens_per_fill: 1000
`

using the following struct provided for the globalRateLimiter Config throws an invalid descriptor error as expeccted.
`      -rate_limit:
             requests_per_unit: 10
             unit: minute  `

Does Envoy not support dynamic values for LocalRateLimiter descriptors .. ??

[optional *Relevant Links*:]
>https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/local_rate_limit_filter#using-rate-limit-descriptors-for-local-rate-limiting

**Comment by htuch** [comment]
@junr03 @mattklein123

**Comment by kiran-bjn** [comment]
Hi @junr03  @mattklein123 , can U answer this please.

**Comment by mattklein123** [comment]
This is basically the same issue as https://github.com/envoyproxy/envoy/issues/19685. I think we should probably also support the substitution formatter for dynamic descriptor values. cc @kyessenov

**Comment by kyessenov** [comment]
I think you can use the expression rate limit action extension for dynamic values. We can add substitution formatter, too, as another extension.

We should document this better. I could only find some test case here https://github.com/istio/proxy/blob/master/test/envoye2e/ratelimit/ratelimit_test.go#L41.

**Comment by sseetharaman6** [comment]
I have a similar issue - I want to be able to set a default rate limit for each client and configure special rate limits for some clients, using local rate limit + descriptors.  I am trying to using rate limit descriptors as in [this example](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/local_rate_limit_filter#using-rate-limit-descriptors-for-local-rate-limiting).   
I am able to populate the relevant client info in the request header using lua filter ,  however the [descriptor entries](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/common/ratelimit/v3/ratelimit.proto.html?highlight=entries#envoy-v3-api-msg-extensions-common-ratelimit-v3-ratelimitdescriptor-entry) are throwing me off.
Is there a way to specify that *each unique client* gets `x` tokens with local rate limit, without statically configuring it for each client? I am not sure how to go about setting the value in the descriptor entry for this scenario. For special clients, I can explicitly set the value to match the client info.

**Comment by kyessenov** [comment]
No, there is no way to add dynamic buckets with local rate limiter AFAICT. You might need some global rate limit with quotas for that.

**Comment by ns-asriram** [comment]
I seem to be having a similar issue as @sseetharaman6 here. Only that I am trying to throttle requests based on PATH. 

The URIs are in the form `/{id}/send` where I am looking to assign 'x' tokens for each such "id" PATH. Since the number of ids is huge and not static, it is not feasible to define descriptor entries for each PATH. 

I have raised [#38008](https://github.com/istio/istio/issues/38008) highlighting the configurations that I have tried.

I see that the feature referenced in this issue (#19685 ) is specific to the Network Rate Limiter, I was wondering if someone can provide me with snippets to achieve this use-case.

**Comment by sreekanth-bg** [comment]
@ns-asriram how were you able to achieve this using global rate limits? ie rate limiting based on the {id} which is in the URL?
